internal/server/singleprocess/state/prune.go

```diff
@@ -0,0 +1,85 @@
+package state
+
+import (
+	"sync"
+
+	"github.com/hashicorp/go-memdb"
+)
+
+type pruneOp struct {
+	lock         *sync.Mutex
+	table, index string
+	indexArgs    []interface{}
+	max          int
+	cur          *int
+	check        func(val interface{}) bool
+}
+
+// pruneOld uses the types in op to scan the table indicated and prune old records.
+// The op's check function can allow the process to skip records that shouldn't be
+// pruned regardless of their age.
+func pruneOld(memTxn *memdb.Txn, op pruneOp) (int, error) {
```

Comment:
I know you test the callers of this, but since this is such an important func and a really genuinely nice abstraction, I'd feel better and recommend unit testing this specific func too.

Reply:
Great call, found a bug while testing `pruneOld` in isolation.